autogen/agentchat/conversable_agent.py

```diff
@@ -300,20 +303,35 @@ def _append_oai_message(self, message: Union[Dict, str], role, conversation_id:
         """
         message = self._message_to_dict(message)
         # create oai message to be appended to the oai conversation that can be passed to oai directly.
-        oai_message = {k: message[k] for k in ("content", "function_call", "name", "context") if k in message}
+        oai_message = {
+            k: message[k]
+            for k in ("content", "function_call", "tool_calls", "tool_responses", "tool_call_id", "name", "context")
+            if k in message and message[k] is not None
+        }
         if "content" not in oai_message:
-            if "function_call" in oai_message:
+            if "function_call" in oai_message or "tool_calls" in oai_message:
                 oai_message["content"] = None  # if only function_call is provided, content will be set to None.
             else:
                 return False
 
-        oai_message["role"] = "function" if message.get("role") == "function" else role
-        if "function_call" in oai_message:
+        for tool_call in oai_message.get("tool_calls", []):
+            function = tool_call.get("function", {})
+            if function.get("name", False):
+                function["name"] = self._normalize_name(function["name"])
```

Comment:
This seems to be a data validation issue. We should move this validation to when the function is added to the llm_config and function map instead of dynamically changing the name of the function.

Reply:
This is actually the tool name as requested by the llm. Multiple times I had the llm request a tool that didn't exist or a bad request for a tool that did exist but in a format that did not pass api validation. Once the name is sanitized, then standard error message that the tool doesn't exist allows the conversation to move forward as expected and the llm fixes its tool call in the next message.

Could probably be done in `generate_oai_reply` instead, `_append_oai_message` seemed like where most validation was being done.

Also good to validate the configured tool name, can add that to.